app/services/theme_settings_migrations_runner.rb

```diff
@@ -0,0 +1,174 @@
+        original_name: migration.original_name,
+        theme_field_id: migration.theme_field_id,
+        settings_before: current_settings,
+        settings_after: migrated_settings,
+      }
+      current_settings = migrated_settings
+      current_migration_version = migration.version
+      results
+    rescue DiscourseJsProcessor::TranspileError => error
+      raise_error(
+        "themes.import_error.migrations.syntax_error",
+        name: migration.original_name,
+        error: error.message,
+      )
+    rescue MiniRacer::V8OutOfMemoryError
+      raise_error(
+        "themes.import_error.migrations.exceeded_memory_limit",
+        name: migration.original_name,
+      )
+    rescue MiniRacer::ScriptTerminatedError
+      raise_error("themes.import_error.migrations.timed_out", name: migration.original_name)
+    rescue MiniRacer::RuntimeError => error
+      message = error.message
+      if message.include?("no_exported_migration_function")
+        raise_error(
+          "themes.import_error.migrations.no_exported_function",
+          name: migration.original_name,
+        )
+      elsif message.include?("default_export_is_not_a_function")
+        raise_error(
+          "themes.import_error.migrations.default_export_not_a_function",
+          name: migration.original_name,
+        )
+      elsif message.include?("migration_function_no_returned_value")
+        raise_error(
+          "themes.import_error.migrations.no_returned_value",
+          name: migration.original_name,
+        )
+      elsif message.include?("migration_function_wrong_return_type")
+        raise_error(
+          "themes.import_error.migrations.wrong_return_type",
+          name: migration.original_name,
+        )
+      else
+        raise_error(
+          "themes.import_error.migrations.runtime_error",
+          name: migration.original_name,
+          error: message,
+        )
+      end
+    end
+  end
+
+  private
+
+  def lookup_pending_migrations_fields
+    @theme
+      .migration_fields
+      .left_joins(:theme_settings_migration)
+      .where(theme_settings_migration: { id: nil })
+  end
+
+  def convert_fields_to_migrations(fields)
+    fields.map do |field|
+      match_data = /\A(?<version>\d{4})-(?<name>.+)/.match(field.name)
+
+      if !match_data
+        raise_error("themes.import_error.migrations.invalid_filename", filename: field.name)
+      end
+
+      version = match_data[:version].to_i
+      name = match_data[:name]
+      Migration.new(version: version, name: name, code: field.value, theme_field_id: field.id)
+    end
+  end
+
+  def lookup_overriden_settings
+    hash = {}
+    @theme.theme_settings.each do |row|
+      type_object = ThemeSettingsManager.create(row.name, nil, row.data_type, @theme)
+      hash[row.name] = type_object.value
+    end
+    hash
+  end
+
+  def execute(migration, settings)
+    context = MiniRacer::Context.new(timeout: @timeout, max_memory: @memory)
+
+    context.load(
```

Comment:
Tiny thing - MiniRacer support passing a `filename: "blah"` kwarg to the `.eval` function. Passing something there for each eval makes it much easier to read any potential error backtraces.

So I'd recommend replacing this `.load` with an `eval(File.read(), filename: "loader.js")`, and also adding a `filename:` to the other eval calls below.

(also, might be worth caching the `File.read` in memory rather than pulling from the filesystem every time we execute a migration)